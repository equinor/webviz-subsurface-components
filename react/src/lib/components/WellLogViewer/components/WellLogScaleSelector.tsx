import React, { Component } from "react";

import WellLogViewer from "../WellLogViewer";
//import SyncLogViewer from "../SyncLogViewer";

import ScaleSelector from "./ScaleSelector";

interface Props {
    parent: /*SyncLogViewer |*/ WellLogViewer;
    label?: string;
    values?: number[];
}
interface State {
    scale: number; // value for scale combo
}

export class WellLogScaleSelector extends Component<Props, State> {
    constructor(props: Props, state: State) {
        super(props, state);

        this.state = {
            scale: 1.0,
        };

        this.onScaleChange = this.onScaleChange.bind(this);
        this.onContentRescale = this.onContentRescale.bind(this);
        this.props.parent.callbacksManager.registerCallback(
            "onContentRescale",
            this.onContentRescale
        );
    }
    componentWillUnmount(): void {
        this.props.parent.callbacksManager.unregisterCallback(
            "onContentRescale",
            this.onContentRescale
        );
    }

    // callback function from Vertical Scale combobox
    onScaleChange(value: number): void {
        const controller = this.props.parent.callbacksManager.controller;
        if (!controller) return;
        controller.setContentScale(value);
    }

    onContentRescale(): void {
        this.setState((state: Readonly<State>) => {
            const controller = this.props.parent.callbacksManager.controller;
            if (!controller) return null;
            const scale = controller.getContentScale();
            if (Math.abs(state.scale - scale) < 1) return null;
            return {
                scale: scale,
            };
        });
    }

    render(): JSX.Element {
        return (
            <div style={{ paddingLeft: "10px", display: "flex" }}>
                {this.props.label && <span>{this.props.label}</span>}
                <span style={{ paddingLeft: "10px" }}>
                    <ScaleSelector
                        onScaleChange={this.onScaleChange}
                        values={this.props.values}
                        scale={this.state.scale}
                    />
                </span>
            </div>
        );
    }
}

export default WellLogScaleSelector;
