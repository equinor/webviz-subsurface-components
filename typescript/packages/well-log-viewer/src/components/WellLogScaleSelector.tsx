import React, { Component } from "react";

import ScaleSelector from "./ScaleSelector";

import type { CallbackManager } from "./CallbackManager";

interface Props {
    callbackManager: CallbackManager | undefined;

    label?: string | JSX.Element;
    values?: number[]; // Available scale values array
    round?: boolean | number; // round the value to a "good" number
}
interface State {
    value: number; // value for scale combo
}

export class WellLogScaleSelector extends Component<Props, State> {
    constructor(props: Props) {
        super(props);

        this.state = {
            value: 1.0,
        };

        this.onChange = this.onChange.bind(this);
        this.onContentRescale = this.onContentRescale.bind(this);
    }

    registerCallbacks(callbackManager: CallbackManager | undefined): void {
        callbackManager?.registerCallback(
            "onContentRescale",
            this.onContentRescale
        );
    }

    unregisterCallbacks(callbackManager: CallbackManager | undefined): void {
        callbackManager?.unregisterCallback(
            "onContentRescale",
            this.onContentRescale
        );
    }

    componentDidMount(): void {
        this.registerCallbacks(this.props.callbackManager);
    }

    componentWillUnmount(): void {
        this.unregisterCallbacks(this.props.callbackManager);
    }

    componentDidUpdate(prevProps: Props): void {
        if (prevProps.callbackManager !== this.props.callbackManager) {
            this.unregisterCallbacks(prevProps.callbackManager);
            this.registerCallbacks(this.props.callbackManager);
        }
    }

    // callback function from Vertical Scale combobox
    onChange(value: number): void {
        const controller = this.props.callbackManager?.controller;
        if (!controller) return;
        controller.setContentScale(value);
    }

    onContentRescale(): void {
        this.setState((state: Readonly<State>) => {
            const controller = this.props.callbackManager?.controller;
            if (!controller) return null;
            const value = controller.getContentScale();
            if (Math.abs(state.value - value) < 1) return null;
            return {
                value: value,
            };
        });
    }

    render(): JSX.Element {
        return (
            <div className="scale">
                {this.props.label && (
                    <span className="scale-label">{this.props.label}</span>
                )}
                <span className="scale-value">
                    <ScaleSelector
                        onChange={this.onChange}
                        values={this.props.values}
                        value={this.state.value}
                        round={this.props.round}
                    />
                </span>
            </div>
        );
    }
}

export default WellLogScaleSelector;
