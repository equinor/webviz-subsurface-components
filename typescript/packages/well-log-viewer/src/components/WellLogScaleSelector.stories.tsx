import type { Meta, StoryObj } from "@storybook/react";
import { expect, userEvent, within } from "@storybook/test";
import React, { useState } from "react";

import type { CallbackManager } from "./CallbackManager";
import WellLogScaleSelector from "./WellLogScaleSelector";

/** Mock CallbackManager for Storybook */
class MockCallbackManager {
    controller = {
        setContentScale: (value: number): void => {
            console.log("Scale set to:", value);
        },
        getContentScale: (): number => {
            return 1.0;
        },
    };
    private callbacks: Map<string, (() => void)[]> = new Map();

    registerCallback(eventName: string, callback: () => void): void {
        if (!this.callbacks.has(eventName)) {
            this.callbacks.set(eventName, []);
        }
        this.callbacks.get(eventName)?.push(callback);
    }

    unregisterCallback(eventName: string, callback: () => void): void {
        const callbacks = this.callbacks.get(eventName);
        if (callbacks) {
            const index = callbacks.indexOf(callback);
            if (index > -1) {
                callbacks.splice(index, 1);
            }
        }
    }

    triggerCallback(eventName: string): void {
        const callbacks = this.callbacks.get(eventName);
        if (callbacks) {
            callbacks.forEach((callback) => callback());
        }
    }
}

const stories: Meta = {
    component: WellLogScaleSelector,
    title: "WellLogViewer/Components/WellLogScaleSelector",
    tags: ["no-screenshot-test"],
    parameters: {
        docs: {
            description: {
                component:
                    "A scale selector component for WellLogViewer that allows users to change the content scale/zoom level. Integrates with CallbackManager to communicate scale changes.",
            },
        },
    },
    argTypes: {
        callbackManager: {
            description: "CallbackManager instance to handle scale changes",
            control: false,
        },
        label: {
            description: "Label to display next to the scale selector",
            control: { type: "text" },
        },
        values: {
            description: "Available scale values array",
            control: { type: "object" },
        },
        round: {
            description:
                'Round the value to a "good" number (true for auto or number for rounding step)',
            control: { type: "boolean" },
        },
    },
};

export default stories;

interface TemplateProps {
    label?: string | JSX.Element;
    values?: number[];
    round?: boolean | number;
}

const Template = (args: TemplateProps) => {
    const [mockManager] = useState(() => new MockCallbackManager());
    const [info, setInfo] = useState("Current scale: 1.0");
    const [currentScaleIndex, setCurrentScaleIndex] = useState(0);

    // Default scale values to cycle through
    const defaultScaleValues = [0.5, 1, 2, 4, 8];

    // Create a wrapper that updates our info display
    React.useEffect(() => {
        const originalSetScale = mockManager.controller.setContentScale;
        mockManager.controller.setContentScale = function (value: number) {
            originalSetScale.call(this, value);
            setInfo(`Current scale: ${value}`);
        };
    }, [mockManager]);

    const handleExternalScaleChange = () => {
        const scaleValues = args.values || defaultScaleValues;
        const nextIndex = (currentScaleIndex + 1) % scaleValues.length;
        const newScale = scaleValues[nextIndex];

        // Update the mock's internal scale
        mockManager.controller.getContentScale = () => newScale;
        mockManager.controller.setContentScale(newScale);

        // Trigger the callback to notify the component
        mockManager.triggerCallback("onContentRescale");

        setCurrentScaleIndex(nextIndex);
    };

    return (
        <div style={{ padding: "20px", fontFamily: "Arial, sans-serif" }}>
            <div style={{ marginBottom: "10px" }}>
                <WellLogScaleSelector
                    callbackManager={mockManager as unknown as CallbackManager}
                    label={args.label}
                    values={args.values}
                    round={args.round}
                />
            </div>
            <div
                style={{
                    padding: "10px",
                    backgroundColor: "#f5f5f5",
                    borderRadius: "4px",
                    fontSize: "14px",
                    color: "#666",
                }}
            >
                {info}
            </div>
            <div style={{ marginTop: "20px" }}>
                <button
                    onClick={handleExternalScaleChange}
                    style={{
                        padding: "8px 16px",
                        backgroundColor: "#007acc",
                        color: "white",
                        border: "none",
                        borderRadius: "4px",
                        cursor: "pointer",
                    }}
                >
                    Simulate External Scale Change
                </button>
            </div>
        </div>
    );
};

export const Default: StoryObj<typeof Template> = {
    args: {
        label: "Zoom Level:",
        round: true,
    },
    render: (args) => <Template {...args} />,
    play: async ({ canvasElement }) => {
        const canvas = within(canvasElement);

        // Wait for component to render
        await new Promise((resolve) => setTimeout(resolve, 100));

        // Test external scale change button
        const externalButton = canvas.getByText(
            "Simulate External Scale Change"
        );
        await userEvent.click(externalButton);

        // Check that the info display updated
        await expect(
            canvas.getByText(/Current scale: 0\.5/)
        ).toBeInTheDocument();

        // Click again to cycle to next value
        await userEvent.click(externalButton);
        await expect(canvas.getByText(/Current scale: 2/)).toBeInTheDocument();
    },
};

export const WithJSXLabel: StoryObj<typeof Template> = {
    args: {
        label: <strong style={{ color: "#007acc" }}>Custom Scale:</strong>,
        round: true,
    },
    render: (args) => <Template {...args} />,
    play: async ({ canvasElement }) => {
        const canvas = within(canvasElement);

        // Verify the custom JSX label is rendered
        await expect(canvas.getByText("Custom Scale:")).toBeInTheDocument();

        // Test interaction
        const externalButton = canvas.getByText(
            "Simulate External Scale Change"
        );
        await userEvent.click(externalButton);
        await expect(
            canvas.getByText(/Current scale: 0\.5/)
        ).toBeInTheDocument();
    },
};

export const WithCustomValues: StoryObj<typeof Template> = {
    args: {
        label: "Scale:",
        values: [0.5, 1, 2, 4, 8, 16, 32],
        round: false,
    },
    render: (args) => <Template {...args} />,
    play: async ({ canvasElement }) => {
        const canvas = within(canvasElement);

        // Test cycling through custom values
        const externalButton = canvas.getByText(
            "Simulate External Scale Change"
        );

        // Start at scale 1, click to go to next value (2)
        await userEvent.click(externalButton);
        await expect(canvas.getByText(/Current scale: 2/)).toBeInTheDocument();

        // Click again to go to next value (4)
        await userEvent.click(externalButton);
        await expect(canvas.getByText(/Current scale: 4/)).toBeInTheDocument();

        // Click multiple times to test cycling back to beginning
        for (let i = 0; i < 5; i++) {
            await userEvent.click(externalButton);
        }
        await expect(
            canvas.getByText(/Current scale: 0\.5/)
        ).toBeInTheDocument();
    },
};

export const WithoutLabel: StoryObj<typeof Template> = {
    args: {
        round: true,
    },
    render: (args) => <Template {...args} />,
    play: async ({ canvasElement }) => {
        const canvas = within(canvasElement);

        // Test that the component works without a label
        const externalButton = canvas.getByText(
            "Simulate External Scale Change"
        );
        await userEvent.click(externalButton);
        await expect(
            canvas.getByText(/Current scale: 0\.5/)
        ).toBeInTheDocument();
    },
};

export const WithRoundingStep: StoryObj<typeof Template> = {
    args: {
        label: "Scale:",
        round: 0.1, // Round to nearest 0.1
        values: [0.01, 0.11, 0.12, 100.01],
    },
    render: (args) => <Template {...args} />,
    play: async ({ canvasElement }) => {
        const canvas = within(canvasElement);

        // Test rounding behavior with custom values
        const externalButton = canvas.getByText(
            "Simulate External Scale Change"
        );

        // Click to cycle through values and verify they appear
        await userEvent.click(externalButton);
        await expect(
            canvas.getByText(/Current scale: 0\.11/)
        ).toBeInTheDocument();

        await userEvent.click(externalButton);
        await expect(
            canvas.getByText(/Current scale: 0\.12/)
        ).toBeInTheDocument();

        await userEvent.click(externalButton);
        await expect(
            canvas.getByText(/Current scale: 100\.01/)
        ).toBeInTheDocument();
    },
};
