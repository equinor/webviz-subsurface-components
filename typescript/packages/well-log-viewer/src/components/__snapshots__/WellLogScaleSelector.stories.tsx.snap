// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WellLogViewer/Components/WellLogScaleSelector Default play-test: welllogviewer-components-welllogscaleselector--default.dom.html 1`] = `
<div style="padding: 20px; font-family: Arial, sans-serif;">
  <div style="margin-bottom: 10px;">
    <div class="scale">
      <span class="scale-label">
        Zoom Level:
      </span>
      <span class="scale-value">
        <select>
          <option value="1">
            1:1
          </option>
          <option value="100">
            1:100
          </option>
          <option value="200">
            1:200
          </option>
          <option value="500">
            1:500
          </option>
          <option value="1000">
            1:1000
          </option>
          <option value="2000">
            1:2000
          </option>
          <option value="5000">
            1:5000
          </option>
          <option value="10000">
            1:10000
          </option>
          <option value="20000">
            1:20000
          </option>
          <option value="50000">
            1:50000
          </option>
        </select>
      </span>
    </div>
  </div>
  <div style="padding: 10px; background-color: rgb(245, 245, 245); border-radius: 4px; font-size: 14px; color: rgb(102, 102, 102);">
    Current scale: 1
  </div>
  <div style="margin-top: 20px;">
    <button style="padding: 8px 16px; background-color: rgb(0, 122, 204); color: white; border: none; border-radius: 4px; cursor: pointer;">
      Simulate External Scale Change
    </button>
  </div>
</div>
`;

exports[`WellLogViewer/Components/WellLogScaleSelector WithCustomValues play-test: welllogviewer-components-welllogscaleselector--with-custom-values.dom.html 1`] = `
<div style="padding: 20px; font-family: Arial, sans-serif;">
  <div style="margin-bottom: 10px;">
    <div class="scale">
      <span class="scale-label">
        Scale:
      </span>
      <span class="scale-value">
        <select>
          <option value="0.5">
            1:0.5
          </option>
          <option value="1">
            1:1
          </option>
          <option value="2">
            1:2
          </option>
          <option value="4">
            1:4
          </option>
          <option value="8">
            1:8
          </option>
          <option value="16">
            1:16
          </option>
          <option value="32">
            1:32
          </option>
        </select>
      </span>
    </div>
  </div>
  <div style="padding: 10px; background-color: rgb(245, 245, 245); border-radius: 4px; font-size: 14px; color: rgb(102, 102, 102);">
    Current scale: 1
  </div>
  <div style="margin-top: 20px;">
    <button style="padding: 8px 16px; background-color: rgb(0, 122, 204); color: white; border: none; border-radius: 4px; cursor: pointer;">
      Simulate External Scale Change
    </button>
  </div>
</div>
`;

exports[`WellLogViewer/Components/WellLogScaleSelector WithJSXLabel play-test: welllogviewer-components-welllogscaleselector--with-jsx-label.dom.html 1`] = `
<div style="padding: 20px; font-family: Arial, sans-serif;">
  <div style="margin-bottom: 10px;">
    <div class="scale">
      <span class="scale-label">
        <strong style="color: rgb(0, 122, 204);">
          Custom Scale:
        </strong>
      </span>
      <span class="scale-value">
        <select>
          <option value="1">
            1:1
          </option>
          <option value="100">
            1:100
          </option>
          <option value="200">
            1:200
          </option>
          <option value="500">
            1:500
          </option>
          <option value="1000">
            1:1000
          </option>
          <option value="2000">
            1:2000
          </option>
          <option value="5000">
            1:5000
          </option>
          <option value="10000">
            1:10000
          </option>
          <option value="20000">
            1:20000
          </option>
          <option value="50000">
            1:50000
          </option>
        </select>
      </span>
    </div>
  </div>
  <div style="padding: 10px; background-color: rgb(245, 245, 245); border-radius: 4px; font-size: 14px; color: rgb(102, 102, 102);">
    Current scale: 1
  </div>
  <div style="margin-top: 20px;">
    <button style="padding: 8px 16px; background-color: rgb(0, 122, 204); color: white; border: none; border-radius: 4px; cursor: pointer;">
      Simulate External Scale Change
    </button>
  </div>
</div>
`;

exports[`WellLogViewer/Components/WellLogScaleSelector WithRoundingStep play-test: welllogviewer-components-welllogscaleselector--with-rounding-step.dom.html 1`] = `
<div style="padding: 20px; font-family: Arial, sans-serif;">
  <div style="margin-bottom: 10px;">
    <div class="scale">
      <span class="scale-label">
        Scale:
      </span>
      <span class="scale-value">
        <select>
          <option value="100">
            1:100
          </option>
          <option value="0.01">
            1:0.01
          </option>
          <option value="0.11">
            1:0.11
          </option>
          <option value="0.12">
            1:0.12
          </option>
          <option value="100.01">
            1:100.01
          </option>
        </select>
      </span>
    </div>
  </div>
  <div style="padding: 10px; background-color: rgb(245, 245, 245); border-radius: 4px; font-size: 14px; color: rgb(102, 102, 102);">
    Current scale: 100.01
  </div>
  <div style="margin-top: 20px;">
    <button style="padding: 8px 16px; background-color: rgb(0, 122, 204); color: white; border: none; border-radius: 4px; cursor: pointer;">
      Simulate External Scale Change
    </button>
  </div>
</div>
`;

exports[`WellLogViewer/Components/WellLogScaleSelector WithoutLabel play-test: welllogviewer-components-welllogscaleselector--without-label.dom.html 1`] = `
<div style="padding: 20px; font-family: Arial, sans-serif;">
  <div style="margin-bottom: 10px;">
    <div class="scale">
      <span class="scale-value">
        <select>
          <option value="1">
            1:1
          </option>
          <option value="100">
            1:100
          </option>
          <option value="200">
            1:200
          </option>
          <option value="500">
            1:500
          </option>
          <option value="1000">
            1:1000
          </option>
          <option value="2000">
            1:2000
          </option>
          <option value="5000">
            1:5000
          </option>
          <option value="10000">
            1:10000
          </option>
          <option value="20000">
            1:20000
          </option>
          <option value="50000">
            1:50000
          </option>
        </select>
      </span>
    </div>
  </div>
  <div style="padding: 10px; background-color: rgb(245, 245, 245); border-radius: 4px; font-size: 14px; color: rgb(102, 102, 102);">
    Current scale: 1
  </div>
  <div style="margin-top: 20px;">
    <button style="padding: 8px 16px; background-color: rgb(0, 122, 204); color: white; border: none; border-radius: 4px; cursor: pointer;">
      Simulate External Scale Change
    </button>
  </div>
</div>
`;
